metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -173,99 +326,150 @@ def __init__(self, config: BigQueryV2Config, report: BigQueryV2Report):
         self.config: BigQueryV2Config = config
         self.report: BigQueryV2Report = report
 
-    def add_config_to_report(self):
-        self.report.query_log_delay = self.config.usage.query_log_delay
-
     def _is_table_allowed(self, table_ref: Optional[BigQueryTableRef]) -> bool:
         return (
             table_ref is not None
             and self.config.dataset_pattern.allowed(table_ref.table_identifier.dataset)
             and self.config.table_pattern.allowed(table_ref.table_identifier.table)
         )
 
-    def generate_usage_for_project(
-        self, project_id: str, tables: Dict[str, List[str]]
+    def run(
+        self, projects: Iterable[str], table_refs: Collection[str]
     ) -> Iterable[MetadataWorkUnit]:
-        aggregated_info: Dict[
-            datetime, Dict[BigQueryTableRef, AggregatedDataset]
-        ] = collections.defaultdict(dict)
+        events = itertools.chain.from_iterable(
+            self._get_usage_events(project) for project in projects
+        )
+        yield from self._run(events, table_refs)
 
-        parsed_events: Iterable[Union[ReadEvent, QueryEvent]]
-        with PerfTimer() as timer:
+    def _run(
+        self, events: Iterable[AuditEvent], table_refs: Collection[str]
+    ) -> Iterable[MetadataWorkUnit]:
+        try:
+            with BigQueryUsageState(self.config) as usage_state:
+                self._ingest_events(events, table_refs, usage_state)
+
+                if self.config.usage.include_operational_stats:
+                    yield from self._generate_operational_workunits(usage_state)
```

Comment:
Although we already filter by table_refs using `ReadEvent.resource` when storing events, to avoid emitting operations for ghost entities, I think, we should add this check within `_generate_operational_workunits`  for `destination_table` as it may be different from above (`query_event.destinationTable`).

Reply:
Hmm, I think we were missing this check in the original code. @treff7es can you confirm this should be a safe filter? Is there any reason not to do the filter when storing events?

Reply:
Discussed in standup, should be fine to filter